contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,350 @@
+     - `--allow-experimental-upgrades`: Allow the user to upgrade to alpha, beta and release candidate versions.
+     - `--allow-release-candidate-upgrades`: Allow the user to upgrade to release candidate versions.
+     - `--print-config`: Print out the configuration read from the file or ConfigMap in the cluster.
+     - `--skip-preflight-checks`: Skip the preflight checks that are run before execution.
+ - `kubeadm upgrade plan`: Serves as a way to show users:
+   - If the cluster is in an upgradeable state.
+   - What versions the user can upgrade the cluster to.
+ - `kubeadm upgrade apply [version]`:
+   - Is the command that actually performs the upgrade (of the control plane).
+   - Will interactively ask the user for a `[Y/n]` confirmation if the upgrade really should be performed.
+   - Relevant flags:
+     - `--dry-run`: Only log the changes that would be made to the cluster without changing any state.
+     - `--force`, `-f`: Ignore any skippable version policy errors and the interactive yes/no upgrade confirmation.
+     - `--image-pull-timeout`: The time to wait for the new control plane images to pull before timing out. Default: 15 minutes.
+     - `--yes`, `-y`: Run the command non-interactively, don't ask for an upgrade confirmation.
+
+## Example user experience
+
+UX when upgrading a patch version:
+
+```console
+$ kubeadm upgrade plan
+[kubeadm upgrade plan output]
+```
+
+To upgrade the kubelets in the cluster, please upgrade them from the source you installed them from (most often using your package manager).
+
+```console
+$ kubeadm upgrade apply --version v1.8.3
+[kubeadm upgrade apply output]
+```
+
+## Upgrade planning
+
+In order for the user to know whether the cluster may be upgraded or not; the `kubeadm upgrade plan` command was written.
+The command:
+
+1. Checks if the cluster is healthy
+  - Does the API server `/healthz` endpoint return `ok`?
+  - Are all nodes in the cluster in the `Ready` state?
+  - (If Static Pods are used) Do all required Static Pod manifests exist on disk?
+  - (If cluster is self-hosted) Are all control plane DaemonSets healthy?
+1. Builds an internal configuration object to apply to the cluster
+  - The `--config` file takes precedence; otherwise the current configuration stored in a ConfigMap in the cluster is used.
+  - The `KubernetesVersion` field is set to the desired version.
+  - The built configuration object is optionally printed to STDOUT if `--print-config` is `true`.
+1. Information about what releases are available are downloaded from the Kubernetes Release & CI GCS buckets.
+  - The `dl.k8s.io/release/stable.txt` endpoint (and optionally `.../stable-1.X.txt`) is queried for the latest stable version to display.
+  - The API Server, kubeadm CLI, and all kubelets' versions are taken into account when computing the possible upgrades.
+  - The command may show zero to four upgrade possibilities.
```

Comment:
is there a reason that four is the upper bound?

Reply:
added a rationale